tensornetwork/tensornetwork.py

```diff
@@ -317,10 +319,32 @@ def update_axis(self, old_axis: int, old_node: Node, new_axis: int,
                        "node1: '{}', axis1: {}, node2: '{}', axis2: {}".format(
                            self, old_node, old_axis, self.node1, self.axis1,
                            self.node2, self.axis2))
-
-  def is_dangling(self):
+  @property
+  def node1(self) -> Node:
+    val = self._node1()
+    if val is None:
+      raise ValueError("node1 for edge '{}' no longer exists.".format(self))
+    return val
+
+  @property
+  def node2(self) -> Optional[Node]:
+    if self._is_dangling:
+      return None
+    if self._node2() is None:
+      raise ValueError("node2 for edge '{}' no longer exists.".format(self))
+    return self._node2()
+  
+  @node1.setter
+  def node1(self, node: Node) -> None:
+    self._node1 = weakref.ref(node)
+
+  @node2.setter
+  def node2(self, node: Optional[Node]) -> None:
+    self._node2 = weakref.ref(node) if node else None
```

Comment:
If `node` is `None` we should probably set `_is_dangling`, too.

Reply:
Good point.